internal/aks/create.go

```diff
@@ -0,0 +1,217 @@
+package aks
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2020-11-01/containerservice"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-10-01/resources"
+	"github.com/Azure/go-autorest/autorest/to"
+	aksv1 "github.com/rancher/aks-operator/pkg/apis/aks.cattle.io/v1"
+)
+
+func CreateResourceGroup(ctx context.Context, groupsClient *resources.GroupsClient, spec *aksv1.AKSClusterConfigSpec) error {
+	_, err := groupsClient.CreateOrUpdate(
+		ctx,
+		spec.ResourceGroup,
+		resources.Group{
+			Name:     to.StringPtr(spec.ResourceGroup),
+			Location: to.StringPtr(spec.ResourceLocation),
+		})
+
+	return err
+}
+
+// createOrUpdateCluster creates a new managed Kubernetes cluster
+func CreateOrUpdateCluster(ctx context.Context, cred *Credentials, clusterClient *containerservice.ManagedClustersClient, spec *aksv1.AKSClusterConfigSpec) (c containerservice.ManagedCluster, err error) {
+
+	dnsPrefix := spec.DNSPrefix
+	if dnsPrefix == nil {
+		dnsPrefix = to.StringPtr(spec.ClusterName)
+	}
+
+	tags := make(map[string]*string)
+	for key, val := range spec.Tags {
+		if val != "" {
+			tags[key] = to.StringPtr(val)
+		}
+	}
+	clusterName := spec.ClusterName
+	tags["displayName"] = to.StringPtr(clusterName)
+
+	kubernetesVersion := spec.KubernetesVersion
+
+	var vmNetSubnetID *string
+	networkProfile := &containerservice.NetworkProfile{}
+	if hasCustomVirtualNetwork(spec) {
+		virtualNetworkResourceGroup := spec.ResourceGroup
+
+		//if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
+		if spec.VirtualNetworkResourceGroup != nil {
+			virtualNetworkResourceGroup = *spec.VirtualNetworkResourceGroup
+		}
+
+		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
+			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
+			spec.SubscriptionID,
+			virtualNetworkResourceGroup,
+			spec.VirtualNetwork,
+			spec.Subnet,
+		))
+
+		networkProfile.DNSServiceIP = spec.NetworkDNSServiceIP
+		networkProfile.DockerBridgeCidr = spec.NetworkDockerBridgeCIDR
+		networkProfile.ServiceCidr = spec.NetworkServiceCIDR
+
+		if spec.NetworkPlugin != nil {
+			networkProfile.NetworkPlugin = containerservice.NetworkPlugin(*spec.NetworkPlugin)
+		} else {
+			networkProfile.NetworkPlugin = containerservice.Kubenet
+		}
+
+		// if network plugin is 'Azure', set PodCIDR
+		if networkProfile.NetworkPlugin == containerservice.Azure {
+			networkProfile.PodCidr = spec.NetworkPodCIDR
+		}
+
+		if spec.LoadBalancerSKU != nil {
+			loadBalancerSku := containerservice.LoadBalancerSku(*spec.LoadBalancerSKU)
+			networkProfile.LoadBalancerSku = loadBalancerSku
+		}
+
+		if spec.NetworkPolicy != nil {
+			networkProfile.NetworkPolicy = containerservice.NetworkPolicy(*spec.NetworkPolicy)
+		}
+	}
+
+	var agentPoolProfiles []containerservice.ManagedClusterAgentPoolProfile
+	if hasAgentPoolProfile(spec) {
```

Comment:
This could be removed, iterating over an empty list is a noop

Reply:
ok